rclcpp/src/rclcpp/executors/executor_entities_collector.cpp

```diff
@@ -0,0 +1,403 @@
+{
+  std::vector<rclcpp::CallbackGroup::WeakPtr> groups;
+  for (const auto & group_ptr : manually_added_groups_) {
+    groups.push_back(group_ptr);
+  }
+  return groups;
+}
+
+std::vector<rclcpp::CallbackGroup::WeakPtr>
+ExecutorEntitiesCollector::get_automatically_added_callback_groups()
+{
+  std::vector<rclcpp::CallbackGroup::WeakPtr> groups;
+  for (auto const & group_ptr : automatically_added_groups_) {
+    groups.push_back(group_ptr);
+  }
+  return groups;
+}
+
+void
+ExecutorEntitiesCollector::update_collections()
+{
+  this->process_queues();
+  this->add_automatically_associated_callback_groups(this->weak_nodes_);
+  this->prune_invalid_nodes_and_groups();
+}
+
+ExecutorEntitiesCollector::NodeCollection::iterator
+ExecutorEntitiesCollector::remove_weak_node(NodeCollection::iterator weak_node)
+{
+  // Disassociate the guard condition from the executor notify waitable
+  auto guard_condition_it = weak_nodes_to_guard_conditions_.find(*weak_node);
+  if (guard_condition_it != weak_nodes_to_guard_conditions_.end()) {
+    this->notify_waitable_->remove_guard_condition(guard_condition_it->second);
+    weak_nodes_to_guard_conditions_.erase(guard_condition_it);
+  }
+
+  // Mark the node as disassociated (if the node is still valid)
+  auto node_ptr = weak_node->lock();
+  if (node_ptr) {
+    std::atomic_bool & has_executor = node_ptr->get_associated_with_executor_atomic();
+    has_executor.store(false);
+  }
+
+  // Remove the node from tracked nodes
+  return weak_nodes_.erase(weak_node);
+}
+
+ExecutorEntitiesCollector::CallbackGroupCollection::iterator
+ExecutorEntitiesCollector::remove_weak_callback_group(
+  CallbackGroupCollection::iterator weak_group_it,
+  CallbackGroupCollection & collection
+)
+{
+  // Disassociate the guard condition from the executor notify waitable
+  auto guard_condition_it = weak_groups_to_guard_conditions_.find(*weak_group_it);
+  if (guard_condition_it != weak_groups_to_guard_conditions_.end()) {
+    this->notify_waitable_->remove_guard_condition(guard_condition_it->second);
+    weak_groups_to_guard_conditions_.erase(guard_condition_it);
+  }
+
+  // Mark the node as disassociated (if the group is still valid)
+  auto group_ptr = weak_group_it->lock();
+
+  if (group_ptr) {
+    if (!group_ptr->has_valid_node()) {
+      throw std::runtime_error("Node must not be deleted before its callback group(s).");
+    }
+    std::atomic_bool & has_executor = group_ptr->get_associated_with_executor_atomic();
+    has_executor.store(false);
+  }
+
+  // Remove the node from tracked nodes
+  return collection.erase(weak_group_it);
+}
+
+void
+ExecutorEntitiesCollector::add_callback_group_to_collection(
+  rclcpp::CallbackGroup::SharedPtr group_ptr,
+  CallbackGroupCollection & collection)
+{
+  auto iter = collection.insert(group_ptr);
+  if (iter.second == false) {
+    throw std::runtime_error("Callback group has already been added to this executor.");
```

Comment:
could the same problem we discussed before (https://github.com/ros2/rclcpp/pull/2143#discussion_r1156370153) apply also here?

I'm still having issues when adding a node

Reply:
I don't think so?  This should on happen in the case that a group has been added twice to the same collector.

Reply:
I reproduced, looking into it.

Reply:
[0a9c9a6](https://github.com/ros2/rclcpp/pull/2143/commits/0a9c9a6403efa6271401e6440b6b425c538e1bda)